### PR TITLE
feat(git): initialise submodules when cloning repos

### LINKED ISSUE #1356: [Feature Request] Add an option to install git submodules before installing dependencies
This is a follow up of this tweet: https://twitter.com/emmenko/status/950122765394366464

### Background
Sometimes you might need to define some dependencies as a git submodule, meaning you checkout the repository through `git submodule` and point the dependency in `package.json` to a local path:

```json
"dependencies": {
  "foo-bar": "file:./my-local-version-of-foo-bar"
}
```

### When would you do something like this?
You might need to do this when you want to install a custom version of a library (e.g. from a fork) and the repository does not contain the package folder on the root level (e.g. with a monorepo setup).
Obviously you want to point npm to the exact location of the package that you want to install but you can't do that if you want to use _github_ as a version url.

```json
"foo-bar": "repo/name#branch
```

So you need to checkout the repository as a submodule, then point the dependency to the correct folder location inside that submodule.

### Proposal
I guess the proposal for this feature would be to have a configuration option to instruct the bot to run `git submodule sync && git submodule update --init` before running `yarn/npm install`.

---
NOTE: I wouldn't consider this a high priority feature and I don't know how many people are actually interested in this. 

Thanks 🙏

### PR DESCRIPTION
**Pull request by JamieMagee**
See https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---recurse-submodulesltpathspec

Fixes #1356

### CODE REVIEW

**Comment by rarkins** [comment]
Question: Don't we need to prevent Renovate from scanning files within the submodules and attempting to update them?

**Comment by JamieMagee** [comment]
Good question, it's not a scenario I've considered. I'll test later.

**Comment by ViceIce** [comment]
Maybe the submodule checkout should be an option, because you have to add an exclude rule for the submodule folders to renovate.

Maybe we can check for a [`.gitmodules`](https://git-scm.com/docs/gitmodules) file and extract the folders to exclude from there?

**Comment by rarkins** [comment]
I think the submodules are only required for artifact generation (lock files, etc), so *potentially* we can lazily fetch them as-needed. But if we fetch them prior to scanning/extracting then we ideally need a way to automatically exclude them. `.gitmodules` could be that way.

**Comment by rarkins** [comment]
There's one possibility I can think of where submodules might be useful during extraction. It's possible for some managers like Maven that a package could point to a "parent" package and that parent package includes information such as the default registry to use. In this case if we don't have submodules resolved then the package package would look like a broken local link. It's an edge case but maybe a valid one.

However, preventing problems during extraction could be pretty simple - we already call `getFileList()` from within gitfs, so it could pretty easily filter out submodule files at the same time.

**Review by rarkins — CHANGES_REQUESTED** [review]
Make sure that the `getFileList()` function doesn't return submodule files to the upper layer which extracts packages

**Comment by JamieMagee** [comment]
Sure, I'll see what the best way to do this is, with an eye on implementing #2502 next.

**Comment by JamieMagee** [comment]
> Make sure that the `getFileList()` function doesn't return submodule files to the upper layer which extracts packages

I've added the check to `getFileList`

> Maybe the submodule checkout should be an option, because you have to add an exclude rule for the submodule folders to renovate.

I'm more of a fan of enabling it by default. I think it's a sensible default, and makes it easier for users to onboard to renovate.

File: lib/platform/git/storage.ts [hunk 1]
```diff
@@ -259,7 +274,12 @@ export class Storage {
     if (!files) {
       return [];
     }
-    return files.split('\n').filter(Boolean);
+    return files
```
Comment (rarkins) [inline lib/platform/git/storage.ts]: Is this cached, or recalculated every time it's called?

Comment (JamieMagee) [reply lib/platform/git/storage.ts]: No caching, it's recalculated every time. I know it's O(n^2), but it's rarely called, and seems to be okay with a large number of submodules (I tested with RedhawkSDR/redhawk which has 53 files, 49 of which are submodules).

**Comment by JamieMagee** [comment]
@rarkins This is ready for review

**Comment by rarkins** [comment]
@JamieMagee hadn't you added code to exclude submodules files from the `getFileList()`? I don't see it now

**Comment by JamieMagee** [comment]
Yeah, I even added tests. Looks like the branch I rebased was out of date with origin, so I pushed an old commit.

**Comment by JamieMagee** [comment]
Sure, I understand that the merging strategy is mostly a personal preference.

**Comment by ajcann** [comment]
@rarkins Are we able to exclude the submodule recurse from happening? Currently this breaks renovate for us as our git submodules belong to private repositories.

**Comment by JamieMagee** [comment]
@ajcann Yeah, this should be fairly easy to add a config flag for.

**Pull request merged**